src/ucs/datastruct/ptr_map.inl

```diff
@@ -0,0 +1,93 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef UCS_PTR_MAP_INL_
+#define UCS_PTR_MAP_INL_
+
+#include "ptr_map.h"
+
+BEGIN_C_DECLS
+
+KHASH_IMPL(ucs_ptr_map_impl, ucs_ptr_map_key_t, void*, 1,
+           kh_int64_hash_func, kh_int64_hash_equal);
+
+static inline ucs_status_t ucs_ptr_map_init(ucs_ptr_map_t *map)
+{
+    map->next_id = 0;
+    kh_init_inplace(ucs_ptr_map_impl, &map->hash);
+    return UCS_OK;
+}
+
+static inline void ucs_ptr_map_destroy(ucs_ptr_map_t *map)
+{
+    kh_destroy_inplace(ucs_ptr_map_impl, &map->hash);
+}
+
+static UCS_F_ALWAYS_INLINE ucs_ptr_map_key_t
+ucs_ptr_map_put(ucs_ptr_map_t *map, void *ptr)
+{
+    ucs_ptr_map_key_t key = (uintptr_t)ptr;
+
+    UCS_STATIC_ASSERT(UCS_PTR_MAP_KEY_INVALID & UCS_PTR_MAP_KEY_ID_FLAG);
+    ucs_assert_always(!(key & UCS_PTR_MAP_KEY_ID_FLAG));
+
+    return key;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_ptr_map_key_t
+ucs_ptr_map_put_safe(ucs_ptr_map_t *map, void *ptr)
+{
+    ucs_ptr_map_key_t key = (map->next_id += 2) | UCS_PTR_MAP_KEY_ID_FLAG;
+    khiter_t iter;
+    int ret;
+
+    iter = kh_put(ucs_ptr_map_impl, &map->hash, key, &ret);
+    if (ucs_unlikely((ret == UCS_KH_PUT_FAILED) ||
+                     (ret == UCS_KH_PUT_KEY_PRESENT))) {
+        return UCS_PTR_MAP_KEY_INVALID;
+    }
+
+    kh_value(&map->hash, iter) = ptr;
+    return key;
+}
+
+static UCS_F_ALWAYS_INLINE void*
+ucs_ptr_map_get(const ucs_ptr_map_t *map, ucs_ptr_map_key_t key)
+{
+    khiter_t iter;
+
+    UCS_STATIC_ASSERT(UCS_PTR_MAP_KEY_INVALID & UCS_PTR_MAP_KEY_ID_FLAG);
+
+    if (ucs_likely(!(key & UCS_PTR_MAP_KEY_ID_FLAG))) {
+        return (void*)key;
+    }
+
+    /* UCS_PTR_MAP_KEY_INVALID is never put, so error input can be checked
+     * here to avoid extra branch */
+    iter = kh_get(ucs_ptr_map_impl, &map->hash, key);
+    return ucs_unlikely(iter == kh_end(&map->hash)) ? NULL :
+           kh_value(&map->hash, iter);
```

Comment:
we can use 'else' since both option are equal in probability

Reply:
why? not present key is unlikely err option

Reply:
i mean vs. line 64 (key & UCS_PTR_MAP_KEY_ID_FLAG)==0 or 1

Reply:
but you suggested to assume direct_ptr usage like fast path

Reply:
ok, you are right, so let's keep it as is
(btw for put(), can also do `if (ucs_unlikely(indirect))`)